### ISSUE TITLE
Filter type "group" does not work when 500+ options are available

### ISSUE THREAD

**Issue by ChVuagniaux** [octobercms/october#3202]
The group filter dosen't work when the Model table defined in `modelClass` have to many row. 

For big application, like a shop by example, it's important to can filter fields with a model that have many records. Example : filter all commands by client from a list of 5000 clients.

##### Actual behavior
1. When open the filter the system do a query for load all records (`select * from table;`) and display them in the list. This behavior is very nice when you have only few items but query a big tables is a big time consumer
2. When the filter is valided, all records are send in the POST request to the server : when you have several thousand of records you have many chances to exceed the PHP max_input_vars limit.

##### Expected behavior
1. Two scope option for can choose **a limit** and an other for can completly **disable** the initial dropdown (before starting to search) can be very useful and lets the AJAX autocomplete do her job (still with a limit).
2. Only send active records to the server on the POST request don't think that is useful to have available fields provided here.

##### Reproduce steps
1. Use this PR https://github.com/daftspunk/oc-test-plugin/pull/44 of the octobrcms test plugin 
2. Run `php artisan test:seed-posts` for seed the table
3. Go to http://octobercms/backend/october/test/posts comment tab & try to filter by post

##### October build
426

**Comment by LukeTowers** [comment]
@ChVuagniaux look at the `recordsPerPage` config option for the list behavior: http://octobercms.com/docs/backend/lists#configuring-list

**Comment by ChVuagniaux** [comment]
Hi @LukeTowers `recordsPerPage` affect the number of item in the list, any impact on the dropdown.

**Comment by LukeTowers** [comment]
Ah, I see. Do you have any suggestions for how this could be resolved?

**Comment by ChVuagniaux** [comment]
@LukeTowers

For part 1 - I suggest to add two params for the list filters `recordsToShow` & `hideIntialDropDown` (names are not very nice but you see what I mean).
```yaml
scopes:
    post:
        label: Post
        modelClass: October\Test\Models\Post
        conditions: id in (:filtered)
        nameFrom: name
        recordsToShow: 10 # Only show 10 items in the list (initial and ajax filtered) -> add limit to select
        
        hideIntialDropDown: true # No show dropdown before starting to type 
              # avoid to wait that the query was done before open the dropdown
              # and no confusions abount only 10 items listed despite the fact that there are 2k in DB
```

For part 2 - Edit JS of dropdown for no longer send availables items on filter request

**Comment by LukeTowers** [comment]
We might want to make the default a little higher than 10, but yes, I see where you're going with it. Is the group filter the only one affected by this?

**Comment by ChVuagniaux** [comment]
Yes, aggree this is not default values default will be more something like `recordsToShow: 50`  & `hideIntialDropDown: false`

Yes, only for group filter

**Comment by LukeTowers** [comment]
Could you make a PR to solve the issue? We'll probably pick something else for the variable names, but that's not super important.

**Comment by pvullioud** [comment]
Hi @LukeTowers 
I just send an email to @daftspunk to offer to sponsor this issue.

**Comment by LukeTowers** [comment]
Thanks @pvullioud! Not sure when he'll be able to respond, he's been under the weather for the past few days.

**Comment by pvullioud** [comment]
no problem. I'll ping them after the release of 426 if I get no reply

**Comment by LarBearrr** [comment]
Any updates on this issue? Seems like mostly a JS fix? My JS is not quite advanced enough to know where to start. I'm looking at [filter.js](https://github.com/octobercms/october/blob/master/modules/system/assets/ui/js/filter.js) and not quite sure what to look for.

**Comment by LukeTowers** [comment]
@pvullioud are you still interested in sponsoring this issue?

**Comment by LukeTowers** [comment]
Perhaps I'll see if I can take a stab at this as part of #4254

**Comment by manearalucaibsell** [comment]
I found another solution that worked on my side:
1. added in config_filter.yaml 3 new options:
modelToFilterInitialListDisplay
modelToFilterInitialListDisplayFIELD
modelToFilterInitialListDisplayActualFIELD

2. added in Filter.php based on those fileds only the values that appear on desired model.

I am not sure is a good fix, but for my case was the only that worked.

I'll present also here my case:
I have a list of tasks (based on a buyed external plugin) and I wanted to filter the tickets based on the last  user's response. My issue was that the user field was trying to display all my 2500 users, but I wanted them to be filtered based only on the users that appear on other different table based on that table id.

So I will try to place here the code changes on my side, just so you can see the changes. 
![2019-11-11_14h50_36](https://user-images.githubusercontent.com/41894290/68588629-b19fd600-0492-11ea-91f7-78e8d5c84c24.png)
![2019-11-11_14h50_56](https://user-images.githubusercontent.com/41894290/68588631-b19fd600-0492-11ea-9f5e-deed60f04122.png)

**Comment by danielbidala** [comment]
Any progress on this? One of my project will go live in few days and need to filter backend orders list by clients (there are more thousand) I'didn't know this limitation :(

**Comment by danielbidala** [comment]
If I search item in the group popover which is outside the 500 record query limit I find it but I get a js error:
![filter_group_js_error](https://user-images.githubusercontent.com/16636699/69250173-76e31f80-0baf-11ea-9451-95c5b02a6d02.PNG)

**Comment by LukeTowers** [comment]
@danielbidala what version of October are you using?

**Comment by danielbidala** [comment]
@LukeTowers I use Build 458.

**Comment by LukeTowers** [comment]
@danielbidala try updating to the develop branch and see if the issue is still present. That sounds like a separate issue from the 500 record limit.

**Comment by danielbidala** [comment]
@LukeTowers I've updated to dev. branch (at least I think so) but the error is there. I get this error when I start typing in popover's search box (even if I have just a few records in my model, not more than 500). I've also disabled all third party js libraries and custom js files, functions.

The filter is working despite the error.

Anyway do you have suggestion how to solve the problem with 500+ records?

**Comment by pvullioud** [comment]
@danielbidala see a solution on my previous comment https://github.com/octobercms/october/issues/3202#issuecomment-539858432

**Comment by danielbidala** [comment]
@pvullioud Thanks for the suggestion, I've seen your solution but I need to filter by multiple choices at the same time.

**Comment by LukeTowers** [comment]
@danielbidala trigger condition is not defined is related to the Trigger API for form fields. Can I see your full filter config and a screenshot of the whole page that you're using the filter on?

**Comment by danielbidala** [comment]
@LukeTowers Thanks for your interest. The js error appears as I begin to type (2nd screenshot below) Here it is:
![screenshot1](https://user-images.githubusercontent.com/16636699/69374918-71690080-0ca7-11ea-80e8-e24a2e8fbfda.PNG)
![screenshot2](https://user-images.githubusercontent.com/16636699/69374943-7cbc2c00-0ca7-11ea-96b8-6b7bd3bd1682.PNG)

```
# ===================================
# Filter Scope Definitions
# ===================================

scopes:
    customer:
        label: Megrendelő
        modelClass: Brightstaff\Brighterp\Models\Client
        conditions: customer_id in (:filtered)
        nameFrom: name
    owner:
        label: Tulajdonos
        modelClass: Brightstaff\Brighterp\Models\Client
        conditions: owner_id in (:filtered)
        nameFrom: name		
    ordered_at:
        label: Dátum
        type: daterange
        conditions: ordered_at >= ':after' AND ordered_at <= ':before'
    order_type:
        label: Típus
        modelClass: Brightstaff\Brighterp\Models\Order
        type: group
        conditions: order_type in (:filtered)
        options: orderTypeOptions
    order_services:
        label: Szolgáltatások
        modelClass: Brightstaff\Brighterp\Models\Order
        type: group
        scope: services
        options: orderServicesOptions
    other_services:
        label: Igények
        modelClass: Brightstaff\Brighterp\Models\Order
        type: group
        scope: otherServices
        options: otherServicesOptions
    status:
        label: Státusz
        modelClass: Brightstaff\Brighterp\Models\Order
        type: group
        conditions: status in (:filtered)
        options: statusOptions
```

**Comment by LukeTowers** [comment]
@danielbidala are you able to provide me a link and credentials to view a live version of the error over PM on Slack?

**Comment by danielbidala** [comment]
@LukeTowers I've sent you the credentials. Thanks in advance.

**Comment by LukeTowers** [comment]
@danielbidala figured out your issue, it was unrelated to this one. Not sure how you're the only one that has triggered it so far though. I'm working on a fix, will be included in build 460.

**Comment by LukeTowers** [comment]
@danielbidala fix here: https://github.com/octobercms/october/commit/106756d656072178feadcbe4a5472dc2821940b0

**Comment by danielbidala** [comment]
@LukeTowers many thanks for your quick fix. I will test it. Is it planned to fix the original issue?

**Comment by LukeTowers** [comment]
@danielbidala probably eventually, but unless someone is willing to sponsor me to take care of it right now or submits a PR to fix the original issue it's not going to happen right away.

**Comment by LukeTowers** [comment]
Reopening in favour of #5140, #4592, & #2602

**Comment by LukeTowers** [comment]
Fixed by https://github.com/octobercms/october/commit/517c588ef7224941af761ba4df4f329dd0a442db

**Comment by msimkunas** [comment]
@LukeTowers The commit you mention doesn't seem to fix the issue completely because this line is still present:

https://github.com/octobercms/october/blob/517c588ef7224941af761ba4df4f329dd0a442db/modules/backend/widgets/Filter.php#L395

**Comment by LukeTowers** [comment]
@msimkunas does removing that line solve the issue?

**Comment by msimkunas** [comment]
It does. However, it seems that it was introduced intentionally because the dropdown couldn't handle more than 500 items properly. I've tested it with ~600 items and it works fine but I don't know how well this scales now that the changes in https://github.com/octobercms/october/commit/517c588ef7224941af761ba4df4f329dd0a442db have been merged.

**Comment by Klaasie** [comment]
@msimkunas I've tested the PR at the time with 1500+ results in the filter box without any issues. See the link to the plugin at the PR https://github.com/octobercms/october/pull/5141 in which a seeder is provided to fill the roles table.

**Comment by LukeTowers** [comment]
Submit a PR to remove that limit if it's not a problem with large numbers of records @Klaasie @msimkunas

**Comment by msimkunas** [comment]
> @msimkunas I've tested the PR at the time with 1500+ results in the filter box without any issues. See the link to the plugin at the PR https://github.com/octobercms/october/pull/5141 in which a seeder is provided to fill the roles table.

The PR in question does not remove the query limit line from Filter.php. Removing it solves my issue.

I'll post a PR soon @LukeTowers
